File: mac/Tests/BG3AssistantTests/BuildImportTests.swift
import XCTest
@testable import BG3HonorAssistant

final class BuildImportTests: XCTestCase {
    func testDraftProducesExplicitLegalStartingSetup() throws {
        let draft = makeDraft(finalSplit: "Monk 6 / Rogue 4 / Fighter 1 / Cleric 1")
        let imported = try draft.importedBuild(sourceURL: URL(string: "https://example.com/build")!)
        let setup = try XCTUnwrap(imported.build.abilitySetups?.first)

        XCTAssertTrue(AbilityProgression.isValidBG3Setup(setup))
        XCTAssertEqual(imported.build.startingAbilityScores?.dexterity, 16)
        XCTAssertEqual(imported.build.startingAbilityScores?.wisdom, 16)
    }

    func testDraftRejectsLevelTwelveClassSplitThatDoesNotTotalTwelve() {
        let draft = makeDraft(finalSplit: "Monk 8 / Rogue 4 / Fighter 1 / Cleric 1")

        XCTAssertThrowsError(try draft.importedBuild(sourceURL: URL(string: "https://example.com/build")!)) { error in
            XCTAssertTrue(error.localizedDescription.contains("totals 14, not 12"))
        }
    }

    func testDraftRepairsInvalidSplitFromCumulativeClassLevels() throws {
        let base = makeDraft(finalSplit: "Monk 8 / Cleric 1 / Rogue 4")
        let rows = [
            (1, "Monk 1"), (2, "Monk 2"), (3, "Monk 3"), (4, "Monk 4"),
            (5, "Monk 5"), (6, "Cleric 1"), (7, "Monk 6"), (8, "Fighter 1"),
            (9, "Rogue 1"), (10, "Rogue 3"), (11, "Rogue 3"), (12, "Rogue 4"),
        ].map { level, take in
            BuildImportLevel(
                level: level,
                take: take,
                subclassChoice: "",
                choices: "",
                tactics: "",
                confidence: "Explicit",
                abilityScoreReset: nil
            )
        }
        let draft = BuildImportDraft(
            name: base.name,
            role: base.role,
            finalSplit: base.finalSplit,
            classProgression: base.classProgression,
            pointBuyScores: base.pointBuyScores,
            bonusTwo: base.bonusTwo,
            bonusOne: base.bonusOne,
            playPattern: base.playPattern,
            caveat: base.caveat,
            levels: rows,
            gear: base.gear
        )

        let imported = try draft.importedBuild(sourceURL: URL(string: "https://example.com/build")!)

        XCTAssertEqual(imported.build.finalSplit, "Monk 6 / Cleric 1 / Fighter 1 / Rogue 4")
    }

    private func makeDraft(finalSplit: String) -> BuildImportDraft {
        BuildImportDraft(
            name: "Open Hand Monk",
            role: "Mobile striker",
            finalSplit: finalSplit,
            classProgression: "Monk, Rogue, Fighter, Cleric",
            pointBuyScores: AbilityScores(
                strength: 10,
                dexterity: 14,
                constitution: 15,
                intelligence: 8,
                wisdom: 15,
                charisma: 8
            ),
            bonusTwo: .dexterity,
            bonusOne: .wisdom,
            playPattern: "Use unarmed attacks and control.",
            caveat: "Requires a respec.",
            levels: [BuildImportLevel(
                level: 12,
                take: "Cleric 1",
                subclassChoice: "War Domain",
                choices: "",
                tactics: "",
                confidence: "Explicit",
                abilityScoreReset: nil
            )],
            gear: []
        )
    }
}

File: windows/tests/BG3HonorAssistant.Core.Tests/Models/BuildImportTests.cs
using BG3HonorAssistant.Core.Models;

namespace BG3HonorAssistant.Core.Tests.Models;

public sealed class BuildImportTests
{
    [Fact]
    public void DraftProducesExplicitLegalStartingSetup()
    {
        var imported = MakeDraft("Monk 6 / Rogue 4 / Fighter 1 / Cleric 1")
            .Import(new Uri("https://example.com/build"));
        var setup = Assert.Single(imported.Build.AbilitySetups!);

        Assert.True(AbilityProgression.IsValidBg3Setup(setup));
        Assert.Equal(16, imported.Build.StartingAbilityScores?.Dexterity);
        Assert.Equal(16, imported.Build.StartingAbilityScores?.Wisdom);
        Assert.Equal(
            "STR 10 / DEX 16 / CON 15 / INT 8 / WIS 16 / CHA 8",
            imported.Build.StartingAbilities);
    }

    [Fact]
    public void DraftRejectsLevelTwelveClassSplitThatDoesNotTotalTwelve()
    {
        var exception = Assert.Throws<BuildImportException>(
            () => MakeDraft("Monk 8 / Rogue 4 / Fighter 1 / Cleric 1")
                .Import(new Uri("https://example.com/build")));

        Assert.Contains("totals 14, not 12", exception.Message);
    }

    [Fact]
    public void DraftRepairsInvalidSplitFromCumulativeClassLevels()
    {
        var source = MakeDraft("Monk 8 / Cleric 1 / Rogue 4");
        var takes = new[]
        {
            "Monk 1",
            "Monk 2",
            "Monk 3",
            "Monk 4",
            "Monk 5",
            "Cleric 1",
            "Monk 6",
            "Fighter 1",
            "Rogue 1",
            "Rogue 3",
            "Rogue 3",
            "Rogue 4",
        };
        var draft = source with
        {
            Levels = takes
                .Select(
                    (take, index) => new BuildImportLevel(
                        index + 1,
                        take,
                        string.Empty,
                        string.Empty,
                        string.Empty,
                        "Explicit"))
                .ToArray(),
        };

        var imported = draft.Import(new Uri("https://example.com/build"));

        Assert.Equal(
            "Monk 6 / Cleric 1 / Fighter 1 / Rogue 4",
            imported.Build.FinalSplit);
    }

    private static BuildImportDraft MakeDraft(string finalSplit) =>
        new()
        {
            Name = "Open Hand Monk",
            Role = "Mobile striker",
            FinalSplit = finalSplit,
            ClassProgression = "Monk, Rogue, Fighter, Cleric",
            PointBuyScores = new AbilityScores(10, 14, 15, 8, 15, 8),
            BonusTwo = Ability.Dexterity,
            BonusOne = Ability.Wisdom,
            PlayPattern = "Use unarmed attacks and control.",
            Caveat = "Requires a respec.",
            Levels =
            [
                new BuildImportLevel(
                    12,
                    "Cleric 1",
                    "War Domain",
                    string.Empty,
                    string.Empty,
                    "Explicit"),
            ],
            Gear = [],
        };
}
